File: backend/app/models/task.py
from sqlalchemy import Column, DateTime, Float, Index, Integer, String, Boolean, func

from app.models import Base


class Task(Base):
    __tablename__ = "tasks"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, nullable=False, index=True)
    title = Column(String(255), nullable=False)
    priority = Column(Integer, nullable=False, default=1)
    energy = Column(Integer, nullable=False, default=3)
    estimated_minutes = Column(Integer, nullable=False, default=0)
    completed = Column(Boolean, default=False, nullable=False)
    completed_at = Column(DateTime, nullable=True)
    created_at = Column(DateTime, default=func.now(), nullable=False)
    abandoned = Column(Boolean, default=False, nullable=False)
    abandoned_at = Column(DateTime, nullable=True)
    latitude = Column(Float, nullable=True)
    longitude = Column(Float, nullable=True)

    __table_args__ = (
        Index("ix_tasks_user_created", "user_id", "created_at"),
        Index("ix_tasks_completed_at", "completed_at"),
    )

File: frontend/lib/models/task.dart
class Task {
  final String id;
  final String title;
  final int priority;
  final int energy;
  final int estimatedMinutes;
  final bool completed;
  final bool abandoned;
  final DateTime? completedAt;
  final DateTime? abandonedAt;
  final double? latitude;
  final double? longitude;

  Task({
    required this.id,
    required this.title,
    required this.priority,
    required this.energy,
    this.estimatedMinutes = 0,
    this.completed = false,
    this.abandoned = false,
    this.completedAt,
    this.abandonedAt,
    this.latitude,
    this.longitude,
  });

  factory Task.fromJson(Map<String, dynamic> json) => Task(
        id: json['id'].toString(),
        title: json['title'] as String,
        priority: json['priority'] as int,
        energy: json['energy'] as int,
        estimatedMinutes: json['estimated_minutes'] as int? ?? 0,
        completed: json['completed'] as bool? ?? false,
        abandoned: json['abandoned'] as bool? ?? false,
        completedAt: json['completed_at'] != null
            ? DateTime.tryParse(json['completed_at'] as String)
            : null,
        abandonedAt: json['abandoned_at'] != null
            ? DateTime.tryParse(json['abandoned_at'] as String)
            : null,
        latitude: (json['latitude'] as num?)?.toDouble(),
        longitude: (json['longitude'] as num?)?.toDouble(),
      );

  Map<String, dynamic> toJson() => {
        'id': id,
        'title': title,
        'priority': priority,
        'energy': energy,
        'estimated_minutes': estimatedMinutes,
        'completed': completed,
        'abandoned': abandoned,
        if (latitude != null) 'latitude': latitude,
        if (longitude != null) 'longitude': longitude,
      };

  Task copyWith({
    String? title,
    int? priority,
    int? energy,
    int? estimatedMinutes,
    bool? completed,
    bool? abandoned,
    DateTime? completedAt,
    DateTime? abandonedAt,
    double? latitude,
    double? longitude,
  }) =>
      Task(
        id: id,
        title: title ?? this.title,
        priority: priority ?? this.priority,
        energy: energy ?? this.energy,
        estimatedMinutes: estimatedMinutes ?? this.estimatedMinutes,
        completed: completed ?? this.completed,
        abandoned: abandoned ?? this.abandoned,
        completedAt: completedAt ?? this.completedAt,
        abandonedAt: abandonedAt ?? this.abandonedAt,
        latitude: latitude ?? this.latitude,
        longitude: longitude ?? this.longitude,
      );
}
